File: java-client/src/main/java/com/couchbase/client/java/datastructures/CouchbaseMap.java
/*
 * Copyright (c) 2019 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.couchbase.client.java.datastructures;

import java.util.AbstractMap;
import java.util.Collections;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;

import com.couchbase.client.core.annotation.Stability;

import com.couchbase.client.core.error.KeyNotFoundException;
import com.couchbase.client.core.msg.kv.SubDocumentOpResponseStatus;
import com.couchbase.client.core.retry.reactor.RetryExhaustedException;
import com.couchbase.client.java.Bucket;

import com.couchbase.client.java.Collection;
import com.couchbase.client.java.json.JsonArray;
import com.couchbase.client.java.json.JsonObject;
import com.couchbase.client.core.error.CASMismatchException;
import com.couchbase.client.core.error.KeyExistsException;
import com.couchbase.client.core.error.subdoc.MultiMutationException;
import com.couchbase.client.core.error.subdoc.PathNotFoundException;
import com.couchbase.client.java.kv.GetOptions;
import com.couchbase.client.java.kv.InsertOptions;
import com.couchbase.client.java.kv.LookupInOptions;
import com.couchbase.client.java.kv.LookupInResult;
import com.couchbase.client.java.kv.LookupInSpec;
import com.couchbase.client.java.kv.MapOptions;
import com.couchbase.client.java.kv.MutateInOptions;
import com.couchbase.client.java.kv.MutateInSpec;
import com.couchbase.client.java.kv.UpsertOptions;

/**
 * A CouchbaseMap is a {@link Map} backed by a {@link Bucket Couchbase} document (more specifically a
 * {@link JsonObject JSON object}).
 *
 * Null keys are NOT permitted, and keys are restricted to {@link String}.
 *
 * Values in a CouchbaseMap are restricted to the types that a {@link JsonObject JSON objects}
 * can contain. JSON sub-objects and sub-arrays can be represented as {@link JsonObject} and {@link JsonArray}
 * respectively.
 *
 * @param <E> the type of values in the map (restricted to {@link JsonObject}.
 *
 * @since 2.3.6
 */

@Stability.Committed
public class CouchbaseMap<E> extends AbstractMap<String, E> {

    private final String id;
    private final Collection collection;
    private final Class<E> entityTypeClass;
    private final MapOptions.Built mapOptions;
    private final GetOptions getOptions;
    private final LookupInOptions lookupInOptions;
    private final InsertOptions insertOptions;
    private final UpsertOptions upsertOptions;
    /**
     * Create a new {@link CouchbaseMap}, backed by the document identified by <code>id</code>
     * in the given Couchbase <code>bucket</code>. Note that if the document already exists,
     * its content will be used as initial content for this collection. Otherwise it is created empty.
     *
     * @param id the id of the Couchbase document to back the map.
     * @param collection the {@link Collection} through which to interact with the document.
     * @param entityType a {@link Class<E>} describing the type of objects used as values in this Map.
     * @param options a {@link MapOptions} to use for all operations on this instance of the map.
     */
    public CouchbaseMap(String id, Collection collection, Class<E> entityType, MapOptions options) {
        this.id = id;
        this.collection = collection;
        this.entityTypeClass = entityType;

        MapOptions.Built optionsIn = options.build();
        MapOptions mapOpts = MapOptions.mapOptions();

        optionsIn.copyInto(mapOpts);
        this.mapOptions = mapOpts.build();
        this.getOptions = optionsIn.getOptions();
        this.lookupInOptions = optionsIn.lookupInOptions();
        this.upsertOptions = optionsIn.upsertOptions();
        this.insertOptions = optionsIn.insertOptions();
    }

    /**
     * Create a new {@link CouchbaseMap}, backed by the document identified by <code>id</code>
     * in the given Couchbase <code>bucket</code>. Note that if the document already exists,
     * its content will be used as initial content for this collection. Otherwise it is created empty.
     *
     * @param id the id of the Couchbase document to back the map.
     * @param collection the {@link Collection} through which to interact with the document.
     * @param entityType a {@link Class<E>} describing the type of objects used as values in this Map.
     */
    public CouchbaseMap(String id, Collection collection, Class<E> entityType) {
        this(id, collection, entityType, MapOptions.mapOptions());
    }

    @Override
    public E put(String key, E value) {
        checkKey(key);

        for(int i = 0; i < mapOptions.casMismatchRetries(); i++) {
            try {
                long returnCas = 0;
                E result = null;
                try {
                    LookupInResult current = collection.lookupIn(id,
                            Collections.singletonList(LookupInSpec.get(key)),
                            lookupInOptions);
                    returnCas = current.cas();
                    if (current.exists(0)) {
                        result = current.contentAs(0, entityTypeClass);
                    }
                } catch (PathNotFoundException e) {
                    // that's ok, we will just upsert anyways, and return null
                } catch (KeyNotFoundException e) {
                    // we will create an empty doc and remember the cas
                    returnCas = createEmpty();
                }
                collection.mutateIn(id,
                        Collections.singletonList(MutateInSpec.upsert(key, value)),
                        mapOptions.mutateInOptions().cas(returnCas));
                return result;
            } catch (CASMismatchException ex) {
                //will need to retry get-and-set
            }
        }
        throw new RetryExhaustedException("Couldn't perform set in less than " + mapOptions.casMismatchRetries() + " iterations.  It is likely concurrent modifications of this document are the reason");
    }

    @Override
    public E get(Object key) {
        String idx = checkKey(key);
        try {
            return collection.lookupIn(id,
                    Collections.singletonList(LookupInSpec.get(idx)),
                    lookupInOptions)
                    .contentAs(0, entityTypeClass);
        } catch (PathNotFoundException e) {
            return null;
        } catch (KeyNotFoundException e) {
            return null;
        }
    }



    @Override
    public E remove(Object key) {
        String idx = checkKey(key);
        for(int i = 0; i < mapOptions.casMismatchRetries(); i++) {
            try {
                LookupInResult current = collection.lookupIn(id,
                        Collections.singletonList(LookupInSpec.get(idx)),
                        lookupInOptions);
                long returnCas = current.cas();
                E result = current.contentAs(0, entityTypeClass);
                collection.mutateIn(id,
                        Collections.singletonList(MutateInSpec.remove(idx)),
                        mapOptions.mutateInOptions().cas(returnCas));
                return result;
            } catch (KeyNotFoundException e) {
                return null;
            } catch (CASMismatchException ex) {
                //will have to retry get-and-remove
            } catch (MultiMutationException ex) {
                if (ex.firstFailureStatus() == SubDocumentOpResponseStatus.PATH_NOT_FOUND) {
                    return null;
                }
                throw ex;
            }
        }
        throw new RetryExhaustedException("Couldn't perform set in less than " + mapOptions.casMismatchRetries() + " iterations.  It is likely concurrent modifications of this document are the reason");
    }

    @Override
    public void clear() {
        //optimized version over AbstractMap's (which uses the entry set)
        collection.remove(id);
    }

    @Override
    public Set<Entry<String, E>> entrySet() {
        JsonObject obj;
        try {
            obj = collection.get(id, getOptions).contentAsObject();
        } catch (KeyNotFoundException e) {
            obj = JsonObject.empty();
        }
        // don't actually create the doc, yet.
        return new CouchbaseEntrySet((Map<String, E> )obj.toMap());
    }

    @Override
    public boolean containsKey(Object key) {
        String idx = checkKey(key);
        try {
            return collection.lookupIn(id,
                    Collections.singletonList(LookupInSpec.exists(idx)),
                    lookupInOptions)
                    .exists(0);
        } catch (KeyNotFoundException e) {
            return false;
        }

    }

    @Override
    public boolean containsValue(Object value) {
        return super.containsValue(value); //TODO use ARRAY_CONTAINS subdoc operator when available
    }

    @Override
    public int size() {
        try {
            LookupInResult current = collection.lookupIn(id,
                    Collections.singletonList(LookupInSpec.count("")),
                    lookupInOptions);
            return current.contentAs(0, Integer.class);
        } catch (KeyNotFoundException e) {
            return 0;
        }
    }

    private String checkKey(Object key) {
        if (key == null) {
            throw new NullPointerException("Unsupported null key");
        }
        return String.valueOf(key);
    }

    private class CouchbaseEntrySet implements Set<Map.Entry<String, E>> {

        private final Set<Map.Entry<String, E>> delegate;

        private CouchbaseEntrySet(Map<String, E> data) {
            this.delegate = data.entrySet();
        }

        @Override
        public int size() {
            return delegate.size();
        }

        @Override
        public boolean isEmpty() {
            return delegate.isEmpty();
        }

        @Override
        public boolean contains(Object o) {
            return delegate.contains(o);
        }

        @Override
        public Iterator<Entry<String, E>> iterator() {
            return new CouchbaseEntrySetIterator(delegate.iterator());
        }

        @Override
        public Object[] toArray() {
            return delegate.toArray();
        }

        @Override
        public <T> T[] toArray(T[] a) {
            return delegate.toArray(a);
        }

        @Override
        public boolean add(Entry<String, E> stringVEntry) {
            return delegate.add(stringVEntry);
        }

        @Override
        public boolean remove(Object o) {
            if (delegate.remove(o)) {
                if (o instanceof Map.Entry) {
                    Entry<String, E> entry = (Entry<String, E>) o;
                    CouchbaseMap.this.remove(entry.getKey());
                } else {
                    throw new IllegalStateException("Expected entrySet remove() to remove an entry");
                }
                return true;
            } else {
                return false;
            }
        }

        @Override
        public boolean containsAll(java.util.Collection<?> c) {
            return delegate.containsAll(c);
        }

        @Override
        public boolean addAll(java.util.Collection<? extends Entry<String, E>> c) {
            return delegate.addAll(c);
        }

        @Override
        public boolean retainAll(java.util.Collection<?> c) {
            return delegate.retainAll(c);
        }

        @Override
        public boolean removeAll(java.util.Collection<?> c) {
            return delegate.removeAll(c);
        }

        @Override
        public void clear() {
            delegate.clear();
            CouchbaseMap.this.clear();
        }

        @Override
        public String toString() {
            return delegate.toString();
        }

        public int hashCode() {
            return delegate.hashCode();
        }

        public boolean equals(Object obj) {
            return delegate.equals(obj);
        }
    }

    private class CouchbaseEntrySetIterator implements Iterator<Entry<String, E>> {

        private final Iterator<Entry<String, E>> delegateItr;
        private Entry<String, E> lastNext = null;

        public CouchbaseEntrySetIterator(Iterator<Entry<String, E>> iterator) {
            this.delegateItr = iterator;
        }

        @Override
        public boolean hasNext() {
            return delegateItr.hasNext();
        }

        @Override
        public Entry<String, E> next() {
            this.lastNext = delegateItr.next();
            return lastNext;
        }

        @Override
        public void remove() {
            if (lastNext == null)
                throw new IllegalStateException("next() hasn't been called before remove()");
            delegateItr.remove();
            CouchbaseMap.this.remove(lastNext.getKey());
        }
    }

    protected long createEmpty() {
        try {
            return collection.insert(id, JsonObject.empty(), insertOptions).cas();
        } catch (KeyExistsException ex) {
            // Ignore concurrent creations, keep on moving.
            // but we need the cas, so...
            return collection.get(id, getOptions).cas();
        }
    }
}

File: scala-client/src/main/scala-2.13+/com/couchbase/client/scala/datastructures/CouchbaseMap.scala
/*
 * Copyright (c) 2019 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.couchbase.client.scala.datastructures

import com.couchbase.client.core.error.subdoc.PathNotFoundException
import com.couchbase.client.core.error.{CasMismatchException, DocumentNotFoundException}
import com.couchbase.client.scala.Collection
import com.couchbase.client.scala.codec.{JsonDeserializer, JsonSerializer}
import com.couchbase.client.scala.json.JsonObjectSafe
import com.couchbase.client.scala.kv._

import scala.collection.mutable
import scala.reflect.runtime.universe._
import scala.util.{Failure, Success, Try}

/** Presents a Scala Map interface on top of a mutable persistent data structure, in the form of a document stored
  * on the cluster.
  */
class CouchbaseMap[T](
    id: String,
    collection: Collection,
    options: Option[CouchbaseCollectionOptions] = None
)(implicit decode: JsonDeserializer[T], encode: JsonSerializer[T], tag: WeakTypeTag[T])
    extends mutable.AbstractMap[String, T] {

  private val Values = "values."

  private val opts: CouchbaseCollectionOptions = options match {
    case Some(v) => v
    case _       => CouchbaseCollectionOptions(collection)
  }

  private val lookupInOptions = LookupInOptions()
    .timeout(opts.timeout)
    .retryStrategy(opts.retryStrategy)
  private val mutateInOptions = MutateInOptions()
    .timeout(opts.timeout)
    .retryStrategy(opts.retryStrategy)
    .durability(opts.durability)
  private val getOptions = GetOptions()
    .timeout(opts.timeout)
    .retryStrategy(opts.retryStrategy)
  private val insertOptions = InsertOptions()
    .timeout(opts.timeout)
    .retryStrategy(opts.retryStrategy)
    .durability(opts.durability)
  private val removeOptions = RemoveOptions()
    .timeout(opts.timeout)
    .retryStrategy(opts.retryStrategy)
    .durability(opts.durability)

  def getInternal(key: String): Try[Option[T]] = {
    val op = collection.lookupIn(
      id,
      Seq(LookupInSpec.get(key)),
      lookupInOptions
    )

    val result: Try[Option[T]] = op.map(result => {
      if (!result.exists(0)) None
      else {
        // The .get will throw on fail, which is what we want
        Some(result.contentAs[T](0).get)
      }
    })

    result
  }

  override def get(key: String): Option[T] = {
    val result = getInternal(key)

    result match {
      case Success(value)                          => value
      case Failure(err: PathNotFoundException)     => None
      case Failure(err: DocumentNotFoundException) => None
      case Failure(err)                            => throw err
    }
  }

  override def apply(key: String): T = {
    val result = getInternal(key)

    result match {
      case Success(Some(value)) => value
      case Success(None) | Failure(_: DocumentNotFoundException) |
          Failure(_: PathNotFoundException) =>
        throw new NoSuchElementException(s"Key $key not found")
      case Failure(err) => throw err
    }
  }

  // A fast version of `remove` that does not have to return the element
  def removeAt(key: String): Unit = {
    collection
      .mutateIn(
        id,
        Seq(MutateInSpec.remove(key)),
        mutateInOptions
      ) match {
      case Success(_)                              =>
      case Failure(err: DocumentNotFoundException) =>
      case Failure(err: PathNotFoundException)     =>
      case Failure(err)                            => throw err
    }
  }

  // Note `removeAt` is more performant as it does not have to return the removed element
  override def remove(key: String): Option[T] = {
    val op = collection.lookupIn(
      id,
      Seq(LookupInSpec.get(key)),
      lookupInOptions
    )

    val result = op.flatMap(result => result.contentAs[T](0))

    result match {
      case Success(value) =>
        val mutateResult = collection.mutateIn(
          id,
          Seq(MutateInSpec.remove(key)),
          mutateInOptions.cas(op.get.cas)
        )

        mutateResult match {
          case Success(_)                         => Some(value)
          case Failure(err: CasMismatchException) =>
            // Recurse to try again
            remove(key)
          case Failure(err) => throw err
        }

      case Failure(err: DocumentNotFoundException) => None
      case Failure(err: PathNotFoundException)     => None
      case Failure(err)                            => throw err
    }
  }

  private def retryIfDocDoesNotExist[_](f: () => Try[_]): Unit = {
    val result: Try[_] = f()

    result match {
      case Success(_) =>
      case Failure(_: DocumentNotFoundException) =>
        initialize()
        retryIfDocDoesNotExist(f)
      case Failure(err) => throw err
    }
  }

  private def initialize(): Unit = {
    // The .get will throw if anything goes wrong
    collection.insert(id, JsonObjectSafe.create).get
  }

  override def size(): Int = {
    val op = collection.lookupIn(
      id,
      Seq(LookupInSpec.count("")),
      lookupInOptions
    )

    val result = op.flatMap(result => result.contentAs[Int](0))

    result match {
      case Success(count)                        => count
      case Failure(_: DocumentNotFoundException) => 0
      case Failure(err)                          => throw err
    }
  }

  private def all(): Map[String, T] = {
    val op = collection.get(id, getOptions)

    val result = op
      .flatMap(_.contentAs[JsonObjectSafe])
      .map(obj => obj.toMap.asInstanceOf[mutable.AnyRefMap[String, T]])

    result match {
      case Success(values: mutable.AnyRefMap[String, T]) => values.toMap
      case Failure(_: DocumentNotFoundException)         => Map.empty[String, T]
      case Failure(err)                                  => throw err
    }
  }

  override def addOne(kv: (String, T)): this.type = {
    val f = () =>
      collection.mutateIn(
        id,
        Seq(MutateInSpec.upsert(kv._1, kv._2)),
        mutateInOptions
      )
    retryIfDocDoesNotExist(f)
    this
  }

  override def subtractOne(key: String): this.type = {
    val result = collection.mutateIn(
      id,
      Seq(MutateInSpec.remove(key)),
      mutateInOptions
    )

    result match {
      case Success(_)                            => this
      case Failure(_: PathNotFoundException)     => this
      case Failure(_: DocumentNotFoundException) => this
      case Failure(err)                          => throw err
    }
  }

  override def iterator: Iterator[(String, T)] = {
    all().iterator
  }
}
